File: 907. Sum of Subarray Minimums/Approach 2 - Monotonic Stack + Dynamic Programming/Sum of Subarray Minimums.js
let sumSubarrayMins = function (arr) {
    const MOD = 1000000007;
    let stack = [];
    
    let dp = new Array(arr.length).fill(0);

    for (let i = 0; i < arr.length; i++) {
        
        while (stack.length && arr[stack.at(-1)] >= arr[i]) {
            stack.pop();
        }

        if (stack.length) {
            const previousSmaller = stack.at(-1);
            dp[i] = dp[previousSmaller] + (i - previousSmaller) * arr[i];
        } else {
            
            dp[i] = (i + 1) * arr[i];
        }
        stack.push(i);
    }

    const sumOfMinimums = dp.reduce((a, b) => a + b, 0);

    return sumOfMinimums % MOD;
};

File: Problems & Solutions/907. Sum of Subarray Minimums/25 November 2022/Approach 2 - Monotonic Stack + Dynamic Programming/Sum of Subarray Minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int MOD = 1000000007;

        Stack<Integer> stack = new Stack<>();

        int[] dp = new int[arr.length];

        for (int i = 0; i < arr.length; i++) {
            while (!stack.empty() && arr[stack.peek()] >= arr[i]) {
                stack.pop();
            }

            if (stack.size() > 0) {
                int previousSmaller = stack.peek();
                dp[i] = dp[previousSmaller] + (i - previousSmaller) * arr[i];
            } else {
                dp[i] = (i + 1) * arr[i];
            }
            // push the current index
            stack.push(i);
        }

        long sumOfMinimums = 0;
        for (int count : dp) {
            sumOfMinimums += count;
            sumOfMinimums %= MOD;
        }

        return (int) sumOfMinimums;
    }
}
